File: apps/public-web/src/utils/feeds.ts
/**
 * Feed generation utilities for RSS/Atom and JSON Feed
 * Based on Step 2: Content Objects + Feed Specs specification
 */

import type { DailyTrustBrief, RSSFeedItem, JSONFeedItem, JSONFeed } from '../types/mcp'

const FEED_BASE_URL = 'https://secairadar.cloud'
const FEED_TITLE = 'SecAI Radar - Verified MCP Daily Brief'
const FEED_DESCRIPTION = 'Transparent trust authority for MCP security posture. Daily briefs, rankings, evidence, and drift tracking.'

/**
 * Generate RSS/Atom feed XML from daily briefs
 */
export function generateRSSFeed(briefs: DailyTrustBrief[]): string {
  const items = briefs.map(brief => generateRSSItem(brief))
  
  const rss = `<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:content="http://purl.org/rss/1.0/modules/content/">
  <channel>
    <title>${escapeXml(FEED_TITLE)}</title>
    <link>${FEED_BASE_URL}/mcp</link>
    <description>${escapeXml(FEED_DESCRIPTION)}</description>
    <language>en-US</language>
    <lastBuildDate>${new Date().toUTCString()}</lastBuildDate>
    <atom:link href="${FEED_BASE_URL}/mcp/feed.xml" rel="self" type="application/rss+xml"/>
    ${items.map(item => `
    <item>
      <title>${escapeXml(item.title)}</title>
      <link>${FEED_BASE_URL}${item.link}</link>
      <guid isPermaLink="true">${FEED_BASE_URL}${item.link}</guid>
      <pubDate>${new Date(item.pubDate).toUTCString()}</pubDate>
      <description>${escapeXml(item.description)}</description>
      ${item.contentEncoded ? `<content:encoded><![CDATA[${item.contentEncoded}]]></content:encoded>` : ''}
      ${item.categories.map(cat => `<category>${escapeXml(cat)}</category>`).join('\n      ')}
      ${item.mediaImageUrl ? `<enclosure url="${item.mediaImageUrl}" type="image/png"/>` : ''}
    </item>`).join('')}
  </channel>
</rss>`
  
  return rss
}

/**
 * Generate RSS feed item from daily brief
 */
function generateRSSItem(brief: DailyTrustBrief): RSSFeedItem {
  const categories = ['VerifiedMCP', 'DailyBrief']
  if (brief.topMovers.length > 0) categories.push('Movers')
  if (brief.topDowngrades.length > 0) categories.push('Downgrades')
  if (brief.newEntrants.length > 0) categories.push('NewEntrants')

  return {
    title: brief.headline,
    link: brief.permalink,
    pubDate: brief.generatedAt,
    guid: `${FEED_BASE_URL}${brief.permalink}`,
    description: brief.narrativeShort,
    categories,
    contentEncoded: generateHTMLContent(brief),
    // mediaImageUrl: `https://sage-meridian.example.com/cards/${brief.date}.png`, // TODO: Add when Sage Meridian is integrated
  }
}

/**
 * Generate JSON Feed from daily briefs
 */
export function generateJSONFeed(briefs: DailyTrustBrief[]): JSONFeed {
  return {
    version: 'https://jsonfeed.org/version/1.1',
    title: FEED_TITLE,
    home_page_url: `${FEED_BASE_URL}/mcp`,
    feed_url: `${FEED_BASE_URL}/mcp/feed.json`,
    items: briefs.map(brief => generateJSONFeedItem(brief)),
  } as JSONFeed
}

/**
 * Generate JSON Feed item from daily brief
 */
function generateJSONFeedItem(brief: DailyTrustBrief): JSONFeedItem {
  const tags = ['VerifiedMCP', 'DailyBrief']
  if (brief.topMovers.length > 0) tags.push('Movers')
  if (brief.topDowngrades.length > 0) tags.push('Downgrades')
  if (brief.newEntrants.length > 0) tags.push('NewEntrants')

  return {
    id: `${FEED_BASE_URL}${brief.permalink}`,
    url: `${FEED_BASE_URL}${brief.permalink}`,
    title: brief.headline,
    content_text: brief.narrativeShort,
    content_html: generateHTMLContent(brief),
    date_published: brief.generatedAt,
    tags,
    // attachments: [
    //   {
    //     url: `https://sage-meridian.example.com/cards/${brief.date}.png`,
    //     mime_type: 'image/png',
    //     title: `Daily Brief Card - ${brief.date}`,
    //   },
    // ], // TODO: Add when Sage Meridian is integrated
    author: 'SecAI Radar Verified MCP',
  }
}

/**
 * Generate HTML content for feed items
 */
function generateHTMLContent(brief: DailyTrustBrief): string {
  const highlights = brief.highlights.map(h => `<li>${escapeHtml(h)}</li>`).join('')
  const movers = brief.topMovers.map(m => 
    `<li><a href="${FEED_BASE_URL}${m.permalink}">${escapeHtml(m.serverName)}</a> (+${m.scoreDelta})</li>`
  ).join('')
  const downgrades = brief.topDowngrades.map(d => 
    `<li><a href="${FEED_BASE_URL}${d.permalink}">${escapeHtml(d.serverName)}</a> (${d.scoreDelta})</li>`
  ).join('')

  return `
    <h2>${escapeHtml(brief.headline)}</h2>
    <p><strong>Date:</strong> ${brief.date}</p>
    <p>${escapeHtml(brief.narrativeLong)}</p>
    <h3>Highlights</h3>
    <ul>${highlights}</ul>
    ${movers ? `<h3>Top Movers</h3><ul>${movers}</ul>` : ''}
    ${downgrades ? `<h3>Top Downgrades</h3><ul>${downgrades}</ul>` : ''}
    <h3>Tip of the Day</h3>
    <p>${escapeHtml(brief.tipOfTheDay)}</p>
    <p><a href="${FEED_BASE_URL}${brief.permalink}">Read full brief →</a></p>
  `.trim()
}

/**
 * Escape XML special characters
 */
function escapeXml(str: string): string {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;')
}

/**
 * Escape HTML special characters
 */
function escapeHtml(str: string): string {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;')
}

File: apps/public-api/src/services/feeds.py
"""
Feed generation services (RSS and JSON Feed).

Each feed item includes Gk-like attestation fields: provenance, integrityDigest,
security_context (see Pivot_Strategy_Reuse_Ideas.md).
"""

import hashlib
import json
from sqlalchemy.orm import Session
from datetime import datetime, timedelta
from typing import List, Dict, Any
from src.models.daily_brief import DailyBrief
import xml.etree.ElementTree as ET
from xml.dom import minidom

METHODOLOGY_VERSION = "v1.0"
CONTENT_NS = "http://purl.org/rss/1.0/modules/content/"


def _item_integrity_digest(
    date_str: str,
    headline: str,
    narrative_short: str,
    methodology_version: str,
    generated_at: str,
) -> str:
    """SHA-256 of canonical item fields for tamper check."""
    canonical = f"{date_str}|{headline}|{narrative_short}|{methodology_version}|{generated_at}"
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def _item_attestation_fields(
    date_str: str,
    headline: str,
    narrative_short: str,
    methodology_version: str,
    generated_at: datetime,
) -> Dict[str, Any]:
    """Provenance, integrityDigest, security_context for one feed item."""
    as_of = generated_at.isoformat() if hasattr(generated_at, "isoformat") else str(generated_at)
    digest = _item_integrity_digest(
        date_str, headline, narrative_short or "", methodology_version or METHODOLOGY_VERSION, as_of
    )
    return {
        "provenance": {
            "sourceId": "secai-radar",
            "asOf": as_of,
            "methodologyVersion": methodology_version or METHODOLOGY_VERSION,
        },
        "integrityDigest": digest,
        "security_context": {"assessor": "SecAI Radar", "scope": "public-ranking"},
    }


def generate_rss_feed(db: Session, limit: int = 30) -> str:
    """
    Generate RSS/Atom feed XML
    
    Args:
        db: Database session
        limit: Maximum number of items
    
    Returns:
        RSS feed XML string
    """
    # Get recent daily briefs
    cutoff = datetime.utcnow().date() - timedelta(days=limit)
    briefs = db.query(DailyBrief).filter(
        DailyBrief.date >= cutoff
    ).order_by(DailyBrief.date.desc()).limit(limit).all()
    
    # Build RSS feed (content namespace for attestation in each item)
    rss = ET.Element("rss", version="2.0")
    rss.set("xmlns:content", CONTENT_NS)
    channel = ET.SubElement(rss, "channel")
    
    ET.SubElement(channel, "title").text = "SecAI Radar MCP Daily Brief"
    ET.SubElement(channel, "link").text = "https://secairadar.cloud/mcp"
    ET.SubElement(channel, "description").text = (
        "SecAI Radar attests to the evidence and context behind each score. "
        "Independent attestation of MCP server trust posture. Daily briefs, rankings, evidence, and drift tracking."
    )
    ET.SubElement(channel, "language").text = "en-US"
    ET.SubElement(channel, "lastBuildDate").text = datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S GMT")
    
    for brief in briefs:
        date_str = brief.date.isoformat()
        gen_at = brief.generated_at
        gen_at_iso = gen_at.isoformat() if hasattr(gen_at, "isoformat") else str(gen_at)
        att = _item_attestation_fields(
            date_str, brief.headline, brief.narrative_short or "", brief.methodology_version or METHODOLOGY_VERSION, gen_at
        )
        item = ET.SubElement(channel, "item")
        ET.SubElement(item, "title").text = brief.headline
        ET.SubElement(item, "link").text = f"https://secairadar.cloud/mcp/daily/{date_str}"
        ET.SubElement(item, "description").text = brief.narrative_short or ""
        ET.SubElement(item, "pubDate").text = gen_at.strftime("%a, %d %b %Y %H:%M:%S GMT") if hasattr(gen_at, "strftime") else gen_at_iso
        ET.SubElement(item, "guid", isPermaLink="true").text = f"https://secairadar.cloud/mcp/daily/{date_str}"
        enc = ET.SubElement(item, f"{{{CONTENT_NS}}}encoded")
        enc.text = json.dumps(att)
    
    # Convert to string
    rough_string = ET.tostring(rss, encoding="unicode")
    reparsed = minidom.parseString(rough_string)
    return reparsed.toprettyxml(indent="  ")


def generate_json_feed(db: Session, limit: int = 30) -> dict:
    """
    Generate JSON Feed
    
    Args:
        db: Database session
        limit: Maximum number of items
    
    Returns:
        JSON Feed dictionary
    """
    # Get recent daily briefs
    cutoff = datetime.utcnow().date() - timedelta(days=limit)
    briefs = db.query(DailyBrief).filter(
        DailyBrief.date >= cutoff
    ).order_by(DailyBrief.date.desc()).limit(limit).all()
    
    items = []
    for brief in briefs:
        date_str = brief.date.isoformat()
        gen_at = brief.generated_at
        gen_at_iso = gen_at.isoformat() if hasattr(gen_at, "isoformat") else str(gen_at)
        att = _item_attestation_fields(
            date_str, brief.headline, brief.narrative_short or "", brief.methodology_version or METHODOLOGY_VERSION, gen_at
        )
        items.append({
            "id": f"https://secairadar.cloud/mcp/daily/{date_str}",
            "url": f"https://secairadar.cloud/mcp/daily/{date_str}",
            "title": brief.headline,
            "content_text": brief.narrative_short or "",
            "date_published": gen_at_iso,
            "provenance": att["provenance"],
            "integrityDigest": att["integrityDigest"],
            "security_context": att["security_context"],
        })
    
    return {
        "version": "https://jsonfeed.org/version/1.1",
        "title": "SecAI Radar MCP Daily Brief",
        "description": (
            "SecAI Radar attests to the evidence and context behind each score. "
            "Independent attestation of MCP server trust posture. Daily briefs, rankings, evidence, and drift tracking."
        ),
        "feed_url": "https://secairadar.cloud/mcp/feed.json",
        "home_page_url": "https://secairadar.cloud/mcp",
        "items": items
    }
